src/pip/_internal/network/auth.py

```diff
@@ -158,6 +159,7 @@ def _get_new_credentials(self, original_url, allow_netrc=True,
         return username, password
 
     def _get_url_and_credentials(self, original_url):
+        # type: (str) -> Tuple[str, Optional[str], Optional[str]]
```

Comment:
AFAICU, this would return a tuple of (`url`, `username`, `password`). How do `username` and `password` are optional? Could it be because of an empty string?

Reply:
This happens because of two different lines of code in this module.

https://github.com/pypa/pip/blob/76b865155ee646927e838c377b46643b51ccd2fd/src/pip/_internal/network/auth.py#L81

where `AuthInfo` is 

https://github.com/pypa/pip/blob/76b865155ee646927e838c377b46643b51ccd2fd/src/pip/_internal/vcs/versioncontrol.py#L37

and https://github.com/pypa/pip/blob/76b865155ee646927e838c377b46643b51ccd2fd/src/pip/_internal/network/auth.py#L173

which assigns `username` and `password` as `None` due to the default get value.

If we change these two lines to 
```
self.passwords = {}  # type: Dict[str, Tuple[str, str]]
``` 

and 
```
username, password = self.passwords.get(netloc, ('', ''))
```

the return types can be changed to `Tuple[str, str, str]`, but this is not a good idea, especially the second change since we have `username` and `password` None checks on which the logic depends, and that would get affected.

Considering those facts, I will keep the return type as `Tuple[str, Optional[str], Optional[str]]`. Please let me know if my understanding makes sense.